_posts/2023-08-05-datafusion_fast_grouping.md

```diff
@@ -0,0 +1,405 @@
+
+DataFusion `28.0.0` uses the same RawTable and still stores group indexes. The major differences, as shown in Figure 4, are:
+
+
+1. Group values are stored either
+    1. Inline in the `RawTable` (for single columns of primitive types), where the conversion to Row format costs more than its benefit
+    2. In a separate [Rows](https://docs.rs/arrow-row/latest/arrow_row/struct.Row.html) structure with a single contiguous allocation for all groups values, rather than an allocation per group. Accumulators manage the state for all the groups internally, so the code to update intermediate values is a tight type specialized loop. The new [`GroupsAccumulator`](https://github.com/apache/arrow-datafusion/blob/a6dcd943051a083693c352c6b4279156548490a0/datafusion/physical-expr/src/aggregate/groups_accumulator/mod.rs#L66-L75) interface results in highly efficient type accumulator update loops.
+
+
+```
+┌───────────────────────────────────┐     ┌───────────────────────┐
+│ ┌ ─ ─ ─ ─ ─ ┐  ┌─────────────────┐│     │ ┏━━━━━━━━━━━━━━━━━━━┓ │
+│                │                 ││     │ ┃  ┌──────────────┐ ┃ │
+│ │           │  │ ┌ ─ ─ ┐┌─────┐  ││     │ ┃  │┌───────────┐ │ ┃ │
+│                │    X   │  5  │  ││     │ ┃  ││  value1   │ │ ┃ │
+│ │           │  │ ├ ─ ─ ┤├─────┤  ││     │ ┃  │└───────────┘ │ ┃ │
+│                │    Q   │  9  │──┼┼──┐  │ ┃  │     ...      │ ┃ │
+│ │           │  │ ├ ─ ─ ┤├─────┤  ││  └──┼─╋─▶│              │ ┃ │
+│                │   ...  │ ... │  ││     │ ┃  │┌───────────┐ │ ┃ │
+│ │           │  │ ├ ─ ─ ┤├─────┤  ││     │ ┃  ││  valueN   │ │ ┃ │
+│                │    H   │  1  │  ││     │ ┃  │└───────────┘ │ ┃ │
+│ │           │  │ ├ ─ ─ ┤├─────┤  ││     │ ┃  │values: Vec<T>│ ┃ │
+│     Rows       │   ...  │ ... │  ││     │ ┃  └──────────────┘ ┃ │
+│ │           │  │ └ ─ ─ ┘└─────┘  ││     │ ┃                   ┃ │
+│  ─ ─ ─ ─ ─ ─   │                 ││     │ ┃ GroupsAccumulator ┃ │
+│                └─────────────────┘│     │ ┗━━━━━━━━━━━━━━━━━━━┛ │
+│                  Hash Table       │     │                       │
+│                                   │     │          ...          │
+└───────────────────────────────────┘     └───────────────────────┘
+  GroupState                               Accumulators
+
+
+Hash table value stores group_indexes     One  GroupsAccumulator
+and group values.                         per aggregate. Each
+                                          stores the state for
+Group values are stored either inline     *ALL* groups, typically
+in the hash table or in a single          using a native Vec<T>
+allocation using the arrow Row format
+```
+
+**Figure 5**: Hash group operator structure in DataFusion `28.0.0`. Group values are stored either directly in the hash table, or in a single allocation using the arrow Row format. The hash table contains group indexes. A single `GroupsAccumulator` stores the per-aggregate state for _all_ groups.
```

Comment:
Primitive group values are also stored in a single allocation using `Vec<T::Native>`, not directly in the hash table?

https://github.com/apache/arrow-datafusion/blob/63ccd4ab8b5852a7c7928b7d41209c57ef5e1af4/datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs#L88-L89

Reply:
This was a later modification - https://github.com/apache/arrow-datafusion/pull/7043